### ISSUE TITLE
another laser feature

### ISSUE THREAD

**Issue by dodolbysuround** [vlachoudis/bCNC#676]
is there any way on bCNC to sweep at constant speed the overall area and
 throw one laser shoot when flying over parts of the pictures that need to
 be engraved?
 pyrograph plugin works almost like this but instead of variing feed rate,
 I'm thinking about turning on and off the laser.

**Comment by Effer** [comment]
It could be add as option to pyrograph plugin. Sadly I can't test results, so not sure if I'll work on this.

**Comment by dodolbysuround** [comment]
I can be your beta tester (-;

**Comment by Effer** [comment]
Can you describe your setup? Laser model/power work area size. Pictures appreciated :)

**Comment by dodolbysuround** [comment]
I'm using a 12 watts chines laser (like this one but more powerfull)
![image](https://user-images.githubusercontent.com/15923700/31466085-4005c55e-aed6-11e7-8088-dd6a7e473177.png)
on a small custom CNCed milling maching (sorry, I have no picture right kno but I'll take some toomorow. the max work area size is around 300mm (x) 110mm (y)
Here is a picture of what I already did with your updated driller plugin and my laser.
![fullsizerender](https://user-images.githubusercontent.com/15923700/31466881-0bc58920-aed9-11e7-9757-db7bc13a4e66.jpg)
it measures around 10 x 10 mm.

**Comment by dodolbysuround** [comment]
here is the picture of my machine.
![img_4039](https://user-images.githubusercontent.com/15923700/31494864-76112920-af55-11e7-8bb8-eea305fc3f39.JPG)

**Comment by indyucoder** [comment]
Nice, look like it is able to engrave on metal.

**Comment by dodolbysuround** [comment]
yes, I can easily replace the laser by a spindle depending on what I need to do. every screws has been replaced by ball screws

**Comment by Effer** [comment]
Wow, that's a big machine!
So you'll like to sweep forward and back and fire the laser when brightness is under a settable value?
Does the laser have to stop when reaching the point? What if there are more points in line to draw?

**Comment by dodolbysuround** [comment]
arf... not that big actually (-;
I have been thinking quite a lot about all that.
There are two problem I can see with pyrographe plugin used with a laser:
- given the laser is always on, the clearest  area is burnt anyway (not strongly but still) 
- due to acceleration and deceleration, boundary lines are not very neat.
to get reed of these problems, I suggest 2 modifications:
- when traveling over clear areas, G0 should be  used and laser should be OFF (M5)
- when traveling over dark area, G1 should be used with a predefined feed rate that could be adjusted with a parameter in the setting list) and laser should be ON (M3)

In my best dreams, the whole sweeping process should be done at max speed (G0) but I'm afraid the laser wont be powerful enough to burn properly the areas it will have to burn.

I suggest we work only in black and write in a first time. so laser power (Sxxxx) can keep constant during the whole sweeping process. However, this would be interesting to get this value as a parameter is the setting list.

To answer your 2 questions: I don't realy know yet... The thing is, from what I observed, the laser delivers more power when fired by pulse than if it is used to draw a line. Furthermore, depending on the engraves material, it seems the darkness is more homogeneous with several little dots rather than one continuous line. Then I may prefer firing the laser by pulses but if it is quicker and/or easier, for the beginning,  it can be steady ON if there is a long dark area to engrave.

Not sure everything is clear...???

**Comment by dodolbysuround** [comment]
hey @Effer 
I was wondering if you found some time and inspiration to modify the pyrographe plug in?

**Comment by Effer** [comment]
Hi @dodolbysuround , inspiration? maybe. time? not at all. :smile:
Hope I can found some spare time next days. 
Probably I'll just add an option to skip cutting/laser when calculated feed is over a desired value.

**Comment by Effer** [comment]
Sorry @dodolbysuround for late response. I should avoid promise something next time 😭 .
Anyway here a test: if the calculated feed (in relation with image brightness) is over "Skipping feed" the path is skipped or laser is turned off (in laser mode).

![image](https://user-images.githubusercontent.com/1007007/32966925-c40e2e46-cbdb-11e7-953f-1f503e9f4cb6.png)

**Beware I didn't test the g-code on machine**
Unzip this file on plugins folder:  [pyrograph2.zip](https://github.com/vlachoudis/bCNC/files/1483702/pyrograph2.zip)

If you make some test let me know (I've buy a small laser but it will take some time to come from east)

**Comment by dodolbysuround** [comment]
Greaaaaaaaaaaaaaat! I'll try to give it a try in the week end!
but please, help me understanding some point:
- There are 2 differents "laser modes": one in grbl and one in bCNC.  isn't it?
- How do I activate bCNC laser mode? ? I managed to do this but I don't know how, and I'm getting troubles trying to set it back in normal mode.
- I guess this is in tool/config tab but my "CNC button is always grey...
- May I have to restart bCNC for this setting to be taken under acount?

**Comment by Effer** [comment]
* Yes, 2 laser mode. One in the grbl latest firmware that introduce some interesting features (like modulating laser power against the feed rate) and on in bCNC.
* to configure laser mode in bCNC is simple a matter to tick a checkbox in Tool->Config->Laser Cutter:
![image](https://user-images.githubusercontent.com/1007007/32980724-40d46874-cc6c-11e7-91f2-8b4e488b8da6.png)
**Then remember to restart bCNC**

I'm going by memory so I could be wrong but when this option is enable most of the plugins avoid to move Z axis and instead turn laser On/Off with M3/M5
